src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+import UI from 'utils/ui';
+import Events from 'utils/backbone.events';
+import { 
+    createElement, 
+    emptyElement, 
+    prependChild, 
+    nextSibling, 
+    previousSibling
+} from 'utils/dom';
+import button from 'view/controls/components/button';
+import { cloneIcon } from 'view/controls/icons';
+import { MenuItem, RadioMenuItem } from 'view/controls/components/menu/menu-item';
+import { MenuTemplate } from 'view/controls/templates/menu/menu';
+import categoryButton from 'view/controls/components/menu/category-button';
+import { isRtl } from 'utils/language';
+
+export default class Menu extends Events {
+    constructor(_name, _parentMenu, _localization, _template = MenuTemplate) {
+        super();
+        this.name = _name;
+        this.isSubmenu = !!_parentMenu;
+        this.el = createElement(_template(this.isSubmenu, _name));
+        this.topbar = this.el.querySelector(`.jw-${this.name}-topbar`);
+        this.children = {};
+        this.openMenus = [];
+        this.items = [];
+        this.visible = false;
+        this.parentMenu = _parentMenu;
+        this.mainMenu = !this.parentMenu ? this : this.parentMenu.mainMenu;
+        this.closeButton = (this.parentMenu && this.parentMenu.closeButton) || this.createCloseButton(_localization);
+        this.categoryButton = null;
+        if (this.isSubmenu) {
+            if (this.parentMenu.name === 'settings') {
+                this.createCategoryButton(_localization);
+            }
+            this.parentMenu.appendMenu(this);
+        } else {
+            this.ui = addGlobalMenuKeyListener(this);
+        }
+        this.onDocumentClick = this.onDocumentClick.bind(this);
```

Comment:
are these binds explicitly needed with class level funcs? it should use the context of the calling obj right?

Reply:
I took it out and it no longer works. I think it's necessary here since otherwise 'this' will be the document.